src/xmpp/xmpp-im/xmpp_forwarding.cpp

```diff
@@ -0,0 +1,231 @@
+#include "xmpp_client.h"
+#include "xmpp_stream.h"
+#include "xmpp_message.h"
+
+namespace XMPP
+{
+
+static const QString xmlns_forward(QStringLiteral("urn:xmpp:forward:0"));
+static const QString xmlns_delay(QStringLiteral("urn:xmpp:delay"));
+
+//--------------------------------------------------
+// class Forwarding
+//--------------------------------------------------
+
+Forwarding::Forwarding()
+    : type_(ForwardedNone)
+{
+}
+
+Forwarding::Forwarding(const Forwarding &other)
+    : type_(other.type_)
+    , ts_(other.ts_)
+    , msg_(other.msg_)
+{
+}
+
+Forwarding::~Forwarding()
+{
+}
+
+Forwarding & Forwarding::operator=(const Forwarding &from)
+{
+    type_ = from.type_;
+    ts_ = from.ts_;
+    msg_ = from.msg_;
+    return *this;
+}
+
+Forwarding::Type Forwarding::type() const
+{
+    return type_;
+}
+void Forwarding::setType(Type type)
+{
+    if (type_ != type) {
+        type_ = type;
+        if (type == ForwardedNone) {
+            ts_ = QDateTime();
+            msg_ = Message();
+        }
+    }
+}
+
+bool Forwarding::isCarbons() const
+{
+    return (type_ == ForwardedCarbonsSent || type_ == ForwardedCarbonsReceived);
+}
+
+QDateTime Forwarding::timeStamp() const
+{
+    if (!ts_.isNull())
+        return ts_;
+    return msg_.timeStamp();
+}
+
+void Forwarding::setTimeStamp(const QDateTime &ts)
+{
+    ts_ = ts;
+}
+
+Message Forwarding::message() const
+{
+    return msg_;
+}
+
+void Forwarding::setMessage(const Message &msg)
+{
+    msg_ = msg;
+}
+
+bool Forwarding::fromXml(const QDomElement &e, Client *client)
+{
+    if (e.tagName() != QLatin1String("forwarded") || e.attribute(QLatin1String("xmlns")) != xmlns_forward)
+        return false;
+
+    bool correct = false;
+    type_ = Forwarding::ForwardedNone;
+    QDomElement child = e.firstChildElement();
+    while (!child.isNull()) {
+        if (child.tagName() == QLatin1String("message")) {
+            if (client->pushMessage()->processXmlSubscribers(child, client, true))
+                break;
+            Stanza s = client->stream().createStanza(addCorrectNS(child));
+            Message msg;
+            if (msg.fromStanza(s, client->manualTimeZoneOffset(), client->timeZoneOffset())) {
+                if (client->pushMessage()->processMessageSubscribers(msg, true))
+                    break;
+                msg_ = msg;
+                type_ = ForwardedMessage;
+                correct = true;
+            }
+        }
+        else if (child.tagName() == QLatin1String("delay") && child.attribute(QLatin1String("xmlns")) == xmlns_delay) {
+            ts_ = QDateTime::fromString(child.attribute(QLatin1String("stamp")).left(19), Qt::ISODate);
+        }
+        child = child.nextSiblingElement();
+    }
+    return correct;
+}
+
+QDomElement Forwarding::toXml(Stream *stream) const
+{
+    if (type_ == ForwardedNone || !msg_)
+        return QDomElement();
+
+    QDomElement e = stream->doc().createElement(QLatin1String("forwarded"));
```

Comment:
There is no QDomNode::createElement() taking a QLatin1String, so it's useless as you make 2 allocations, one for QLatin1String and one for QString, just use QString::fromLatin1("...") then.